### PR TITLE
Charts with data filter

### PR DESCRIPTION
**Pull request by nickzhihovich**
Add date filter to pie chart

https://drive.google.com/file/d/1W-yKyGEfOpaCY6PyizV5KWi8uZoOrLoD/view?usp=sharing

### CODE REVIEW

File: app/services/charts/home/categories_with_changes.rb [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  private
+
+  def categories_with_changes
+    @categories_with_changes = []
+    categories_that_was_changed.each do |category|
+      item = {category: category, changes: incame_to_category(category)}
+      @categories_with_changes << item
+    end
+    @categories_with_changes
+  end
```
Comment (AleksSenkou) [inline app/services/charts/home/categories_with_changes.rb]: @marderer see `#inject` method

File: app/services/charts/home/categories_with_changes.rb [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+  def categories_that_was_changed
+    @categories = []
+    category_transactions.each do |transaction|
+      @categories << transaction.transactinable.category
+    end
+    @categories = @categories.uniq
```
Comment (AleksSenkou) [inline app/services/charts/home/categories_with_changes.rb]: @marderer here too

File: spec/services/charts/home/categories_with_changes_date_spec.rb [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  let(:categories_titles) { categories.pluck(:title) }
+
+  let(:changes) do
+    Charts::Home::CategoriesWithChanges.new(Transaction.category_transactions).call
+  end
+
+  let(:data_chenges) { described_class.new(changes).date }
+
+  before do
+    BalanceTransactions::Creator.new(balance_params).create
```
Comment (AleksSenkou) [inline spec/services/charts/home/categories_with_changes_date_spec.rb]: @marderer why didn't you use factories for that? Update you factories with such logic if needed

**Review by AleksSenkou — COMMENTED** [review]
SUggestions:

File: app/controllers/pages_controller.rb [hunk 4]
```diff
@@ -1,9 +1,24 @@
...
+    @date = Charts::Home::GetData.new(@transactions).call
+  end
+
+  def date_chart
+  end
+
+  private
+
+  def search_transactions
+    return unless user_signed_in?
```
Comment (AleksSenkou) [inline app/controllers/pages_controller.rb]: @marderer what do you think about removing this line and add `authenticate_user!` action before `#home` too?

```ruby
before_action :authenticate_user!
before_action :search_transactions
```

File: app/controllers/pages_controller.rb [hunk 5]
```diff
@@ -1,9 +1,24 @@
...
+    @search = current_user.transactions.includes(:transactinable).ransack(params[:q])
+    @transactions = @search.result(distinct: true)
+    @data = Charts::Home::GetData.new(@transactions).call
```
Comment (AleksSenkou) [inline app/controllers/pages_controller.rb]: @marderer do you use `@search` or `@transactions` in views?

File: app/services/charts/home/categories_with_changes.rb [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+  def call
+    categories_with_changes
+  end
+
+  private
+
+  def categories_with_changes
+    categories_that_was_changed.reduce([]) do |categories, item|
+      categories << {category: item, changes: incame_to_category(item)}
+    end
```
Comment (AleksSenkou) [inline app/services/charts/home/categories_with_changes.rb]: @marderer what about:
```ruby
categories_that_was_changed.map do |item|
  {category: item, changes: incame_to_category(item)}
end
```

it's the same

File: app/services/charts/home/categories_with_changes.rb [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+  end
+
+  def categories_that_was_changed
+    category_transactions.reduce([]) do |categories, transaction|
+      categories << transaction.transactinable.category
+    end.uniq
+  end
```
Comment (AleksSenkou) [inline app/services/charts/home/categories_with_changes.rb]: @marderer here too

File: app/views/pages/_date_chart.html.slim [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+          right:0,
+          bottom:0,
+          top:0
+        }
+      },
+      tooltips:{
+        enabled:true
+      }
+    }
+  });
```
Comment (AleksSenkou) [inline app/views/pages/_date_chart.html.slim]: @marderer why didn't you use `src/javascript/charts.js`? If you need some data from the server than just use `data-` html attribute and read it via jQuery

File: spec/services/charts/home/categories_with_changes_date_spec.rb [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+  before do
+    create(:balance_transaction_creator, amount: balance, user: user)
+
+    5.times do
+      CategoryTransactions::Creator.new(
+        amount: amount,
+        user_id: user.id,
+        category_id: category.id
+      ).create
+    end
```
Comment (AleksSenkou) [inline spec/services/charts/home/categories_with_changes_date_spec.rb]: @marderer let's go with similar solution with `factory :balance_transaction_creator` and use `create_list`

**Review by AleksSenkou — COMMENTED** [review]
Suggestions

File: app/views/pages/_date_chart.html.slim (hunk 8 above)
Comment (marderer) [reply app/views/pages/_date_chart.html.slim]: @AleksSenkou We should use `date-` attributes. It from trello task.

File: app/controllers/pages_controller.rb (hunk 5 above)
Comment (marderer) [reply app/controllers/pages_controller.rb]: @AleksSenkou no

File: spec/services/charts/home/categories_with_changes_date_spec.rb (hunk 9 above)
Comment (marderer) [reply spec/services/charts/home/categories_with_changes_date_spec.rb]: @AleksSenkou I tried but I didn't succeed. I spent more than a day and did not work, can I later ask you about this and fix it in another PR?

Comment (AleksSenkou) [reply spec/services/charts/home/categories_with_changes_date_spec.rb]: @marderer yeah, we can discuss it later

**Review by AleksSenkou — APPROVED** [review]
LGTM

**Pull request merged**